### PR TITLE
Add Log In/Log Out features

### PR DESCRIPTION
**Pull request by shreyabarua**
Users who are signed up on ResumeBuddy can now log in and out. When a user is logged in, they can also view "My Account".

### CODE REVIEW

File: resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+    String urlToRedirectToAfterUserLogsOut = "/index.html";
+    String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+
+    if (status) {
+      jsonLogin = "{\"status\": true, ";
+
+    } else {
+      jsonLogin = "{\"status\": false, ";
+    }
```
Comment (astrelni) [inline resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java]: Nit: could convert `status` to string via `String.valueOf(status)` or `Boolean.toString(status)`. Then you wouldn't need the `if` / `else`.

File: resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  public boolean validEmail(String queryType, String email_key) {
+    // search through the respective database and make sure that user is signed up on the website
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query(queryType);
+    PreparedQuery results = datastore.prepare(query);
+    for (Entity entity : results.asIterable()) {
+      String email = (String) entity.getProperty("email");
+      if (email.equals(email_key)) return true;
+    }
+    return false;
```
Comment (astrelni) [inline resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java]: I think you can accomplish this also by adding a filter to the query:

```
Filter emailFilter = new FilterPredicate("email", FilterOperator.EQUAL, email_key);
query.setFilter(emailFilter);
```

Then `results` will contain only matching entities so it is enough to check if `results` is not empty:

```
return results.countEntities(FetchOptions.Builder.withDefaults()) != 0;
```

File: resume-buddy/src/main/webapp/script.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+function login() {
+    fetch('/login').then(response => response.json()).then((login) => {
+        const loginLinkElement = document.getElementById('login-container');
+        const myAccountElement = document.getElementById('my-account');
+        if (login.status) {
+            //Show the 'my account' option
+            //Show a log out option
```
Comment (astrelni) [inline resume-buddy/src/main/webapp/script.js]: Nit: add a space between `//` and start of comment text.

File: resume-buddy/src/main/webapp/script.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+            myAccountElement.style.display = "block";
+            loginLinkElement.innerHTML = "<a style=\"text-decoration:none\" href=\"" + 
+            login.logout_url + "\">Log Out</a>";
+            
+        }
+        else {
+            //Show the log in option
```
Comment (astrelni) [inline resume-buddy/src/main/webapp/script.js]: Nit: add a space between `//` and start of comment text.

File: resume-buddy/src/main/webapp/index.html [hunk 5]
```diff
@@ -21,7 +21,8 @@ <h2><b>ResumeBuddy</b></h2>
         <a href="reviewer-form.html" class="w3-bar-item w3-button">Review</a>
         <a href="reviewee-form.html" class="w3-bar-item w3-button">Get a Review</a>
         <a href="#" class="w3-bar-item w3-button">Meet the Team</a>
-        <a href="#" class="w3-bar-item w3-button">Log In</a>
+        <a href="#" class="w3-bar-item w3-button" id = "login-container" onclick="login()">Log In</a>
+        <a href="resume-review.html" class="w3-bar-item w3-button" id = "my-account">My Account</a>
```
Comment (astrelni) [inline resume-buddy/src/main/webapp/index.html]: Nit: inconsistent spacing around `=` for attributes.

File: resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
```
Comment (astrelni) [inline resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java]: Nit: avoid vertical whitespace at the start of the function.

**Review by astrelni — CHANGES_REQUESTED** [review]

File: resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 6 above)
Comment (shreyabarua) [reply resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java]: @astrelni by vertical white space do you mean before 'public'?

Comment (astrelni) [reply resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java]: @shreyabarua sorry I meant the blank line 38 after the `{` and before `UserService ...`. Unless the auto-formatter added this? As far as I'm aware, just about all styles say to not start a function with a blank line.

Comment (shreyabarua) [reply resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java]: I see that makes sense, fixed it now!

File: resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    boolean status;
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java]: Nit: make this variable name more specific. Like `isReviewerOrReviewee`

File: resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+    jsonLogin = "{\"status\": " + String.valueOf(status) + ", ";
+    jsonLogin += "\"login_url\": \"" + loginUrl + "\", ";
+    jsonLogin += "\"logout_url\": \"" + logoutUrl + "\"}";
+    // send the json as the response
+    response.setContentType("application/json;");
+    response.getWriter().println(jsonLogin);
+  }
+
+  public boolean validEmail(String queryType, String email_key) {
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java:61]: Add comment explaining what this function does

**Review by oydang — CHANGES_REQUESTED** [review]

**Review by oydang — APPROVED** [review]

**Pull request merged**